### PR TITLE
add folder data source

### PR DESCRIPTION
**Pull request by fantapop**
This change adds the folder data source and associated tests. There was previously a folder resource only.  The new data source can find a folder using either a path string or the ID of the folder.  For example:

    data "cockroach_folder" "team1" {
      path = "/prod/team1"
    }

    data "cockroach_folder" "prod" {
      id = var.prod_folder_id
    }

**Commit checklist**
- [x] Changelog
- [x] Doc gen (`make generate`)
- [x] Integration test(s)
- [x] Acceptance test(s)
- [x] Example(s)

### CODE REVIEW

File: examples/data-sources/cockroach_folder/data-source.tf [hunk 1]
```diff
@@ -0,0 +1,11 @@
+variable "prod_folder_id" {
+  type = string
+}
+
+data "cockroach_folder" "team1" {
+  path = "/prod/team1"
+}
+
+data "cockroach_folder" "prod" {
```
Comment (pawalt) [inline examples/data-sources/cockroach_folder/data-source.tf:9]: I really like the design of this data source - the pathing is nice and intuitive!

File: internal/provider/folder_data_source.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+			"id": schema.StringAttribute{
+				MarkdownDescription: "The id the folder.",
+				Optional:            true,
+			},
+			"path": schema.StringAttribute{
+				MarkdownDescription: "An absolute path to the folder. Trailing slashes are optional. (i.e. /folder1/folder2)",
+				Optional:            true,
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "Name of the cluster.",
```
Comment (pawalt) [inline internal/provider/folder_data_source.go:56]: s/cluster/folder/

File: internal/provider/folder_data_source.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	var folderDataSource FolderDataSourceModel
+	diags := req.Config.Get(ctx, &folderDataSource)
+
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		resp.Diagnostics.AddWarning("Error loading the folder", "")
+		return
+	}
+
+	if folderDataSource.ID.IsNull() && folderDataSource.Path.IsNull() {
```
Comment (pawalt) [inline internal/provider/folder_data_source.go:103]: Is there a way we can fail earlier here? Maybe on `Configure`?

File: internal/provider/folder_data_source.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+			"At least one of ID or Path must be set, both were empty",
+		)
+		return
+	}
+
+	var folder *client.FolderResource
+	if !folderDataSource.ID.IsNull() {
+		folderID := folderDataSource.ID.ValueString()
+
+		if !uuidRegex.MatchString(folderID) {
```
Comment (pawalt) [inline internal/provider/folder_data_source.go:115]: Interesting that we have this. Maybe we should choose between using the regex and using `uuid.Parse`

File: internal/provider/folder_data_source.go (hunk 2 above)
Comment (fantapop) [reply internal/provider/folder_data_source.go:56]: cool. I also caught a few other copy pasta mistakes.

File: internal/provider/folder_data_source.go (hunk 3 above)
Comment (fantapop) [reply internal/provider/folder_data_source.go:103]: There is a built in validation framework.  I'll look into using that.

File: internal/provider/folder_data_source.go (hunk 4 above)
Comment (fantapop) [reply internal/provider/folder_data_source.go:115]: hmm, yeah.  the mix of these is no bueno.  I'll send a followup PR to replace the regexp with `uuid.Parse`

File: vendor/modules.txt [hunk 5]
```diff
@@ -57,7 +57,7 @@ github.com/cloudflare/circl/math/mlsbset
 github.com/cloudflare/circl/sign
 github.com/cloudflare/circl/sign/ed25519
 github.com/cloudflare/circl/sign/ed448
-# github.com/cockroachdb/cockroach-cloud-sdk-go v1.8.0
+# github.com/cockroachdb/cockroach-cloud-sdk-go v1.8.0 => ../cockroach-cloud-sdk-go
```
Comment (pawalt) [inline vendor/modules.txt:60]: I think this will need to land before this can be safely comitted.

**Review by pawalt — APPROVED** [review]

File: vendor/modules.txt (hunk 5 above)
Comment (fantapop) [reply vendor/modules.txt:60]: yes... I didn't mean to push this change up 🤦

File: internal/provider/folder_data_source.go (hunk 3 above)
Comment (fantapop) [reply internal/provider/folder_data_source.go]: I've incorporated the validation framework.  On one hand its nice to use something built in but on the other the error messages can be pretty bad.   For parent_id the field either needs to be a uuid or match the string "root".  But the framework gives 2 distinct errors when it doesn't match:


```
➜  cockroach_folder git:(CC-27470-support-new-terraform-data-source-for-folders) ✗ terraform plan
╷
│ Error: Invalid Attribute Value Match
│
│   with cockroach_folder.example_folder_child,
│   on main.tf line 59, in resource "cockroach_folder" "example_folder_child":
│   59:   parent_id = "d01989b0-b005-493d-a0ae-585e7fc93b7"
│
│ Attribute parent_id value must be one of: ["root"], got: "d01989b0-b005-493d-a0ae-585e7fc93b7"
╵
╷
│ Error: Invalid Attribute Value Match
│
│   with cockroach_folder.example_folder_child,
│   on main.tf line 59, in resource "cockroach_folder" "example_folder_child":
│   59:   parent_id = "d01989b0-b005-493d-a0ae-585e7fc93b7"
│
│ Attribute parent_id must match UUID format, got: d01989b0-b005-493d-a0ae-585e7fc93b7
╵
```

I may look a bit more for a better option.

Comment (fantapop) [reply internal/provider/folder_data_source.go]: I went ahead and added a custom validator for the parent_id.  It was pretty straightforward and improves the experience.  https://github.com/cockroachdb/terraform-provider-cockroach/pull/190/files#diff-c7ceb2bc425e0c0b8e8ad8826979ce04e42f1151285d04332a9fce92ca30f7b3

File: examples/workflows/cockroach_folder/main.tf [hunk 6]
```diff
@@ -49,40 +42,32 @@ provider "cockroach" {
...
 }
 
-resource "cockroach_folder" "example-folder-child" {
+resource "cockroach_folder" "example_folder_child" {
   name      = var.folder_child_name
-  parent_id = cockroach_folder.example-folder-parent.id
+  parent_id = cockroach_folder.example_folder_parent.id
+}
+
+resource "cockroach_user_role_grant" "folder_admin_grant" {
```
Comment (marksoper) [inline examples/workflows/cockroach_folder/main.tf:55]: This is cool to see @fantapop - tying folders in with user_role_grant

**Pull request merged**